tck/src/main/java/org/eclipse/microprofile/concurrency/tck/ConcurrencyManagerTest.java

```diff
@@ -0,0 +1,93 @@
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.arquillian.testng.Arquillian;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.testng.Assert;
+import org.testng.ITestResult;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+public class ConcurrencyManagerTest extends Arquillian {
+
+    @AfterMethod
+    public void afterMethod(Method m, ITestResult result) {
+        System.out.println("<<< END " + m.getClass().getSimpleName() + '.' + m.getName() + (result.isSuccess() ? " SUCCESS" : " FAILED"));
+        Throwable failure = result.getThrowable();
+        if (failure != null) {
+            failure.printStackTrace(System.out);
+        }
+    }
+
+    @BeforeMethod
+    public void beforeMethod(Method m) {
+        System.out.println(">>> BEGIN " + m.getClass().getSimpleName() + '.' + m.getName());
+    }
+
+    @Deployment
+    public static WebArchive createDeployment() {
+        return ShrinkWrap.create(WebArchive.class, ConcurrencyManagerTest.class.getSimpleName() + ".war")
+                .addClass(ConcurrencyManagerTest.class);
+    }
+
+    /**
+     * Verify obtaining a ConcurrencyManager builder from the ConcurrencyProvider. Then validate building,
+     * registering, and releasing a custom ConcurrencyManager. If a ConcurrencyManager builder is not supported
+     * then this test will no-op.
+     */
+    @Test
+    public void builderForConcurrencyManagerIsProvided() {
+        try {
+            ConcurrencyProvider provider = ConcurrencyProvider.instance();
+            ClassLoader classLoader = ConcurrencyManagerTest.class.getClassLoader();
+            
+            //obtain the ConcurrencyManagerBuilder
+            Builder concurrencyManagerBuilder = provider.getConcurrencyManagerBuilder();
+            Assert.assertNotNull(concurrencyManagerBuilder,
+                    "MicroProfile Concurrency implementation does not provide a ConcurrencyManager builder.");
+            
+            //build and register a ConcurrencyManager
+            ConcurrencyManager builtManager = concurrencyManagerBuilder.build();
+            provider.registerConcurrencyManager(builtManager, classLoader);
+            ConcurrencyManager registeredManager = provider.getConcurrencyManager(classLoader);
+            Assert.assertEquals(builtManager, registeredManager,
+                    "ConcurrencyManager.getConcurrencyManager(classLoader) did not return the same manager that was registered.");
+            
+            //release the ConcurrencyManager
+            provider.releaseConcurrencyManager(registeredManager);
+            Assert.assertNotEquals(builtManager, provider.getConcurrencyManager(classLoader),
+                    "ConcurrencyManager was not released from the ConcurencyProvider.");
+            
+        } 
+        catch (UnsupportedOperationException ex) {
```

Comment:
I think the UnsupportedOperationException should be more granularly placed.  For example, an implementation should not be allowed to support registerConcurrencyManager but reject releaseConcurrencyManager.  Instead, an initial check should be make if concurrencyProvider.getConcurrencyManagerBuilder is supported, and that should set expectations for what the other methods are allowed to raise UnsupportedOperationException for.

Reply:
Updates for UnsupportedOperationException checking have been made in commit 2.